File: gto-support-jsonapi-retrofit2/src/main/java/org/ccci/gto/android/common/jsonapi/retrofit2/BaseJsonApiParams.java
package org.ccci.gto.android.common.jsonapi.retrofit2;

import com.google.common.base.Joiner;
import com.google.common.base.Strings;
import com.google.common.collect.ForwardingMap;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

public abstract class BaseJsonApiParams<T extends BaseJsonApiParams> extends ForwardingMap<String, String> {
    @VisibleForTesting
    static final String PARAM_INCLUDE = "include";
    private static final String PARAM_FIELDS = "fields";
    private static final String PARAM_SORT = "sort";

    private static final Joiner CSV_JOINER = Joiner.on(",").skipNulls();

    private final Map<String, String> mParams = new HashMap<>();

    @Override
    protected Map<String, String> delegate() {
        return mParams;
    }

    @NonNull
    protected abstract T self();

    @NonNull
    public final T addAll(@NonNull final Map<String, String>... params) {
        for (final Map<String, String> param : params) {
            if (param != null) {
                putAll(param);
            }
        }
        return self();
    }

    @NonNull
    public final T include(final String... paths) {
        put(PARAM_INCLUDE, CSV_JOINER.join(Strings.emptyToNull(get(PARAM_INCLUDE)), null, (Object[]) paths));
        return self();
    }

    @NonNull
    public final T clearIncludes() {
        remove(PARAM_INCLUDE);
        return self();
    }

    @NonNull
    public final T fields(@NonNull final String type, @NonNull final String... fields) {
        final String param = PARAM_FIELDS + "[" + type + "]";
        put(param, CSV_JOINER.join(Strings.emptyToNull(get(param)), null, (Object[]) fields));
        return self();
    }

    @NonNull
    public final T clearFields(@NonNull final String type) {
        remove(PARAM_FIELDS + "[" + type + "]");
        return self();
    }

    @NonNull
    public final T sort(@NonNull final String... keys) {
        put(PARAM_SORT, CSV_JOINER.join(keys));
        return self();
    }
}

File: gto-support-jsonapi-retrofit2/src/main/java/org/ccci/gto/android/common/jsonapi/retrofit2/BaseJsonApiParams.kt
package org.ccci.gto.android.common.jsonapi.retrofit2

import androidx.annotation.VisibleForTesting

@VisibleForTesting
internal const val PARAM_INCLUDE = "include"
private const val PARAM_FIELDS = "fields"
private const val PARAM_SORT = "sort"

abstract class BaseJsonApiParams<T : BaseJsonApiParams<T>> : MutableMap<String, String> by mutableMapOf() {
    protected abstract val self: T

    fun addAll(vararg params: Map<String, String>): T {
        params.forEach { putAll(it) }
        return self
    }

    fun include(vararg paths: String): T {
        this[PARAM_INCLUDE] = listOf(this[PARAM_INCLUDE], *paths).filterNot { it.isNullOrEmpty() }.joinToString(",")
        return self
    }

    fun clearIncludes(): T {
        remove(PARAM_INCLUDE)
        return self
    }

    fun fields(type: String, vararg fields: String): T {
        val param = fieldsFor(type)
        this[param] = listOf(this[param], *fields).filterNot { it.isNullOrEmpty() }.joinToString(",")
        return self
    }

    fun clearFields(type: String): T {
        remove(fieldsFor(type))
        return self
    }

    fun sort(vararg keys: String): T {
        put(PARAM_SORT, keys.joinToString(","))
        return self
    }
}

private fun fieldsFor(type: String) = "$PARAM_FIELDS[$type]"
